### PR TITLE
Track total hits up to 10,000 by default

### LINKED ISSUE #33028: Do not compute hit counts by default
**Context**
Lucene 8 introduces optimizations that allow to compute top hits more efficiently by skipping documents that do not produce competitive scores. We would like to enable this behavior by default so that users can opt in if they need accurate total hit counts, which are costly, rather than the other way around.

Not returning a hit count at all is problematic for traditional search UIs with pagination: say that you want to display up to 5 pages with 10 hits per page, you need to know whether the hit count is between 0 and 10, between 11 and 20, between 21 and 30, between 31 and 40, or greater than 40 in order to know how many pages need to be displayed. In order to address this issue, Lucene takes a configurable threshold: if the hit count is less than this threshold, then you will get an accurate hit count, otherwise you will get a lower bound of the hit count.

We don't want to discuss backward compatibility for now, let's focus what we want to have eventually, and only then discuss how we get there to make the change easy to digest for users. That's fine if we need multiple steps and the whole change is only available in 8 rather than 7.

**Response format**
We need a way to tell users whether the hit count is accurate or a lower bound. Multiple ideas have been mentioned:

 0. not modify the response format: if the user asked to count up to X and the hit count is greater than X, just return X as the hit count
 1. use a string that ends with a `+` as a way to say that the hit count is a lower bound, eg. `{ "hits": { "total": "1234+" } }` when the hit count is a lower bound and `{ "hits": { "total": 1234 } }` like today otherwise
 2. use another field that tells whether the hit count is accurate or a lower bound, eg. `{ "hits": { "total": 1234, "total_hits_relation": "gte" } }`
 3. make `hits.total` an object with a value and a relation, eg. `{ "hits": { "total": { "value": 1234, "relation": "gte" } } }`
 4. make `hits.total` an object that has two possible keys but only one is ever set, eg. `{ "hits": { "total": { "gte": 1234 } } }` or `{ "hits": { "total": { "eq": 1234 } } }`
 5. don't reuse `hits.total` at all and return a different field if ie. `hits.min` or some better name.

When we discussed these options, we felt like 1 would make parsing more complicated, and with 2 it would be too easy to miss the fact that you need to look up another field in order to know how to interpret the hit count.

**Implementation options**

**Option 1: Make `track_total_hits` take a number**

We already have a `track_total_hits` switch which we added for index sorting, but currently take a boolean. It would be easy to make it take a number instead that would be the minimum number of hits to count accurately.

We could ease transition to the new response format by using the current format when `track_total_hits` is unset or set to a boolean, and the new format when `track_total_hits` is set to a number, and then deprecate support for booleans.

Users who want accurate hit counts could set a very high value for this parameter, we could potentially allow using a special value like `-1` as a way to mean "be accurate".

**Option 2: Hardcoded number of hits to count**

Always count accurately up to `index.max_result_window` hits. If users need accurate hits, they will need to use an aggregation (we need to add such an aggregation that counts docs).

Ping @elastic/es-clients to get opinions about the above thoughts, especially the response format.

### PR DESCRIPTION
**Pull request by jimczi**
This commit changes the default for the `track_total_hits` option of the search request
to `10,000`. This means that by default search requests will accurately track the total hit count
up to `10,000` documents, requests that match more than this value will set the `"total.relation"`
to `"gte"` (e.g. greater than or equals) and the `"total.value"` to `10,000` in the search response.
Scroll queries are not impacted, they will continue to count the total hits accurately.
The default is set back to `true` (accurate hit count) if `rest_total_hits_as_int` is set in the search request.
I choose `10,000` as the default because that's also the number we use to limit pagination. This means that
users will be able to know how far they can jump (up to 10,000) even if the total number of hits is not accurate.

Closes #33028

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -204,6 +204,13 @@ public int readInt() throws IOException {
                 | ((readByte() & 0xFF) << 8) | (readByte() & 0xFF);
     }
 
+    public Integer readOptionalInt() throws IOException {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Can you add javadocs?

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 2]
```diff
@@ -322,6 +322,15 @@ public void writeOptionalString(@Nullable String str) throws IOException {
         }
     }
 
+    public void writeOptionalInt(@Nullable Integer integer) throws IOException {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: can you add javadocs?

File: server/src/main/java/org/elasticsearch/search/query/EarlyTerminatingCollector.java [hunk 3]
```diff
@@ -78,4 +81,8 @@ public void collect(int doc) throws IOException {
             };
         };
     }
+
+    public boolean hasEarlyTerminated() {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/query/EarlyTerminatingCollector.java]: can you add javadocs?

**Review by jpountz — APPROVED** [review]
The change looks good. We will probably need to talk with other teams before merging this change so that they set `track_total_hits=true` and can deal with disabling exact counts later?

**Comment by jimczi** [comment]
run the gradle build tests 1

**Comment by jimczi** [comment]
@elasticmachine run gradle build tests 1

**Comment by jimczi** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 4]
```diff
@@ -696,6 +696,15 @@ int getNumBuffered() {
         int getNumReducePhases() { return numReducePhases; }
     }
 
+    private int resolveTrackTotalHits(SearchRequest request) {
+        if (request.scroll() != null) {
+            // no matter what the value of track_total_hits is
+            return SearchContext.TRACK_TOTAL_HITS_ACCURATE;
+        }
+        return request.source() == null ? SearchContext.DEFAULT_TRACK_TOTAL_HITS_UP_TO : request.source().trackTotalHitsUpTo() == null ?
```
Comment (adhulipa) [inline server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: is the `request.source() == null?` check necessary on line 704? Seems like it's redundant given the check on line 700

p.s: I'm a curious engineer looking to learn more about elasticsearch's internals. I'm reading the diffs in my spare time to learn how experts in the field are building real-world distributed systems. Thanks for taking the time to answer my questions!

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: The check is on `request.scroll` which is different than `request.source` so it is necessary.

**Pull request merged**